### PR TITLE
Integrate ClamAV virus scanning functionality into OpenProject

### PR DESCRIPTION
**Pull request by oliverguenther**
https://community.openproject.org/projects/openproject/work_packages/52909/activity

# Acceptance criteria

## Administration

- [x] A new attachment menu item is created
    - The previous system settings &gt; Attachments item is being moved to that new item
    - A subitem &quot;Quarantined attachments&quot; is being added below this new menu node
- [x] Attachments setting are extended with Antivirus settings
    - [x] A mode to operate the virus scanning
        - Disabled
        - ClamAV (local socket)
        - ClamAV (remote host)
    - [x] When any option other than Disabled is shown, these additional fields are shown
        - Host name or Socket for the ClamAV connection
        - Action to perform for attachments on which viruses have been found
            - **Quarantine:** Show the file once uploaded, but disallow access to it.
            - **Delete:** Delete the file as soon as the virus was found
- [x] Quarantined attachments shows a flat table of all currently quarantined files, restricted to administrators
    - Newly quarantined files are showing up there
    - Administrators can delete the file or override the virus scanning decision by the system

## Enabling virus scaning

- When the user first enables virus scanning, they might already have uploaded attachments
- [x] Previously uploaded attachments are scanned in a background job, while they remain accessible as before.
    - [x] IF viruses are found for these attachments, they are placed into quarantine or deleted according to the setting.

## Disabling virus scaning

- When the user disables virus scanning, some files might still be in quarantine or in queue for scanning
    - [x] An informational text is shown when disabling virus scanning with remaining quarantined files. The &quot;Quarantined attachments&quot; remain available even when disabled to perform actions on it.
    - [x] Attachment scanning will be skipped as soon as the functionality is disabled.

## Scanning of attachments

- [x] After uploading of an attachment, or after a direct upload (e.g. S3) is finished, a scanning background job is triggered
- [x] If the background job finds a virus, it will perform these actions depending on the selected action setting
    - [x] **Quarantine:** Keep the file, but prevent users from downloading it. Show the quarantined file under the attachment administration.  A journal entry will be added informing users about the quarantining.
    - [x] **Delete:** Delete the file as soon as the virus was found. A journal entry will be added informing users about the removal
- [x] The uploaded file is only accessible to the original author until the scan has completed, allowing them to seamlessly use it in e.g., WYSIWYG
- [x] The uploaded files are shown, but are not yet accessible to other users. An error/warning message is returned when the file is not yet downloadable.
- [x] Quarantined files are shown, but not accessible for download. A journal entry/comment is shown when the file has been quarantined or deleted.

### CODE REVIEW

File: app/views/admin/settings/virus_scanning_settings/show.html.erb [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program; if not, write to the Free Software
+Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+
+See COPYRIGHT and LICENSE files for more details.
+
+++#%>
+<%= toolbar title: t(:'settings.antivirus.title') %>
```
Comment (HDinger) [inline app/views/admin/settings/virus_scanning_settings/show.html.erb]: Is there a reason you did not use the `PageHeader` component here?

File: app/views/admin/attachments/quarantined_attachments/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+<% html_title t('antivirus_scan.quarantined_attachments.title') %>
+
+<%=
+  render(Primer::OpenProject::PageHeader.new) do |header|
```
Comment (HDinger) [inline app/views/admin/attachments/quarantined_attachments/index.html.erb:4]: This does not show a breadcrumb which all other admin pages do.. The PageHeader component has a `with_breadcrumbs` slot that can be used for that. Ideally, we could use the same data from the controller that we used for the breadcrumbs in the old toolbar definition but I am not sure if that is possible. If you want to, we can have a short chat about it.

**Review by HDinger — APPROVED** [review]
Looks good to me 👍 

I just have a minor remark with regards to the PageHeader. Further, I was wondering whether the "Quarantine" page should also be hidden if there is no EE token available. Probably not as the admins should still be able to remove the quarantine of attachments 🤔 Maybe we can shortly discuss that with the prodcut team today.

File: app/views/admin/attachments/quarantined_attachments/index.html.erb (hunk 2 above)
Comment (oliverguenther) [reply app/views/admin/attachments/quarantined_attachments/index.html.erb:4]: Currently the breadcrumbs are rendered separately with the  `show_local_breadcrumb` method which was not set in quarantined controller. I fixed that

**Comment by oliverguenther** [comment]
> Looks good to me 👍
> 
> I just have a minor remark with regards to the PageHeader. Further, I was wondering whether the "Quarantine" page should also be hidden if there is no EE token available. Probably not as the admins should still be able to remove the quarantine of attachments 🤔 Maybe we can shortly discuss that with the prodcut team today.

I've changed the check to show it if EE is enabled OR there are any quarantined files. this would mean the item would vanish once there are no items left but I think that edge case is fine to ignore.

**Pull request merged**